README.md
# inventory

A Clojure test of web service using reitit, http-kit and an atom as database.

## Usage

Clone the project.
```
git clone https://github.com/edurevsky/clojure-inventory.git
```
Go to the project's directory
```
cd clojure-inventory
```
Run with leiningen
```
lein run
```

The server will listen at port 8888

## Routes
- If a request or response doesn't follow the expected
schema, the server will give a json response with a Clojure schema.core error
### /inventory
Get Mapping:
```json lines
// returns
{
  "inventory-items": 0,
  "inventory-value": 0
}
```
### /items
Post Mapping:
```json lines
// Receives
{
  "name": "xpto",
  "price": 2.99,
  "quantity": 2
}
// Returns 201
{
  "id": "9ad71e6b-3c85-4176-bd26-f82957473ccf",
  "name": "xpto",
  "quantity": 234,
  "price": 2.99,
  "total-price": 5.98
}
```
### /items/:id
Get Mapping:
```json lines
// /items/9ad71e6b-3c85-4176-bd26-f82957473ccf
// Returns 200 or 404 if the id is not present
// in the db
{
  "id": "9ad71e6b-3c85-4176-bd26-f82957473ccf",
  "name": "xpto",
  "quantity": 2,
  "price": 2.99,
  "total-price": 5.98
}
```
Put Mapping:
```json lines
// /items/9ad71e6b-3c85-4176-bd26-f82957473ccf
// Receives
{
  "name": "xpto2",
  "price": 3.99
}
// Returns 200 or 404 if the id is not present
// in the db
{
  "id": "9ad71e6b-3c85-4176-bd26-f82957473ccf",
  "name": "xpto2",
  "quantity": 2,
  "price": 3.99,
  "total-price": 7.98
}
```
Delete Mapping:
```json lines
// /items/9ad71e6b-3c85-4176-bd26-f82957473ccf

// Returns 204 on successful delete or 404 
// if the id is not present in the db
```
### /purchase
Put Mapping:
```json lines
// Receives
{
  "item-id": "9ad71e6b-3c85-4176-bd26-f82957473ccf",
  "amount": 2
}
// Returns 204 or 404 if the id is not present
// in the db
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-08-29
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-08-29
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/inventory/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/inventory/compare/0.1.0...0.1.1


test/inventory/core_test.clj
(ns inventory.core-test
  (:require [clojure.test :refer :all]
            [inventory.core :refer :all]
            [muuntaja.core :as m]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 1 1))))

(deftest handler-test
  (testing "Should return status 200."
    (let [request {:request-method :get :uri "/ok"}
          response (handler request)
          expected 200]
      (is (= (:status response) expected))))
  (testing "Should return status 200 and not have an empty body."
    (let [request {:request-method :get :uri "/inventory"}
          response (handler request)
          expected-status 200
          body (m/decode-response-body response)]
      (is
        (and
          (= (:status response) expected-status)
          (contains? body :inventory-items)
          (contains? body :inventory-value)))))
  (testing "Should return status 201 and body id must not be nil."
    (let [request {:request-method :post
                   :uri            "/items"
                   :body-params    {:name     "name"
                                    :quantity 1
                                    :price    0.1}}
          response (handler request)
          expected-status 201
          body (m/decode-response-body response)]
      (is
        (and
          (= (:status response) expected-status)
          (not (nil? (:id body)))))))
  (testing "Should return status 400 when requesting with an invalid body."
    (let [request {:request-method :post
                   :uri            "/items"
                   :body-params    {:quantity 1
                                    :price    0.1}}
          response (handler request)
          expected-status 400]
      (is (= (:status response) expected-status)))))

test/inventory/handlers_test.clj
(ns inventory.handlers-test
  (:require [clojure.test :refer :all]
            [inventory.handlers :refer :all]
            [schema.core :as s]
            [inventory.models :as m]))

(deftest handle-get-item-by-id-test
  (testing "That returns a map with status 404."
    (let [mocked-request {:db          (atom {})
                          :path-params {:id "id"}}
          expected-status 404]
      (is (:status (handle-get-item-by-id mocked-request) expected-status))))
  (testing "That returns a map with status 200 and body that is valid as an Item."
    (let [gen-id (random-uuid)
          mocked-request {:db          (atom {gen-id {:id gen-id :name "name" :price 0.1 :quantity 1}})
                          :path-params {:id (str gen-id)}}
          resp (handle-get-item-by-id mocked-request)
          expected-status 200]
      (is
        (and
          (= (:status resp) expected-status)
          (s/validate m/Item (:body resp)))))))

(deftest handle-delete-item-by-id-test
  (testing "That returns a map with status 404."
    (let [mocked-request {:db          (atom {})
                          :path-params {:id "id"}}
          resp (handle-delete-item-by-id mocked-request)
          expected-status 404]
      (is (= (:status resp) expected-status))))
  (testing "That returns a map with status 204."
    (let [gen-id (random-uuid)
          mocked-request {:db          (atom {gen-id {:id gen-id :name "name" :price 0.1 :quantity 1}})
                          :path-params {:id (str gen-id)}}
          resp (handle-delete-item-by-id mocked-request)
          expected-status 204]
      (is (= (:status resp) expected-status)))))

(deftest handle-item-purchase-test
  (let [id #uuid"b894ffbc-f586-47a2-a7ee-77ad775d96b4"
        item {:id id :name "?" :price 0.01 :quantity 2321}
        initial-value {id item}]
    (testing "That returns a map with status 400 when requesting with an non existing id."
      (let [mocked-request {:db          (atom initial-value)
                            :body-params {:item-id (str (random-uuid)) :amount 1}}
            resp (handle-item-purchase mocked-request)]
        (is (= 400
               (:status resp)))))
    (testing "That returns a map with status 400 when requesting an amount that is higher than the available."
      (let [mocked-request {:db          (atom initial-value)
                            :body-params {:item-id (str id) :amount 2322}}
            resp (handle-item-purchase mocked-request)]
        (is (= 400
               (:status resp)))))
    (testing "That returns a map with status 204 when requesting an amount that is the same as the available."
      (let [mocked-request {:db          (atom initial-value)
                            :body-params {:item-id (str id) :amount 2321}}
            resp (handle-item-purchase mocked-request)]
        (is (= 204
               (:status resp)))))
    (testing "That returns a map with status 204 when requesting an amount that is below the available."
      (let [mocked-request {:db          (atom initial-value)
                            :body-params {:item-id (str id) :amount 2320}}
            resp (handle-item-purchase mocked-request)]
        (is (= 204
               (:status resp))))))
  (testing "That returns a map with status 204."
    (let [id #uuid"fdc8d2eb-ea52-4cf0-81c2-b5392afe822f"
          item {:id       id
                :name     "xpto"
                :price    0.99
                :quantity 10}
          mocked-request {:db          (atom {id item})
                          :body-params {:item-id "fdc8d2eb-ea52-4cf0-81c2-b5392afe822f" :amount 1}}
          resp (handle-item-purchase mocked-request)]
      (is (= 204
             (:status resp))))))

(deftest handle-update-item-by-id-tests
  (testing "That returns a map with status 404."
    (let [request {:db (atom {})
                   :path-params {:id (str (random-uuid))}}
          resp (handle-update-item-by-id request)]
      (is (= 404
             (:status resp)))))
  (testing "That returns a map with status 200 and updates the name and price of the Item."
    (let [id1 #uuid"e749beed-d16f-4d2a-a1a4-4a888a83b90d"
          id2 #uuid"8f2142f2-aeda-4f4f-b5d3-0a1f207d10c3"
          item1 {:id id1 :name "asqfe" :price 9.21 :quantity 28}
          item2 {:id id2 :name "fi2o" :price 4.85 :quantity 1}
          db {id1 item1 id2 item2}
          req1 {:db (atom db)
                :path-params {:id "e749beed-d16f-4d2a-a1a4-4a888a83b90d"}
                :body-params {:name "pppp" :price 23.12}}
          req2 {:db (atom db)
                :path-params {:id "8f2142f2-aeda-4f4f-b5d3-0a1f207d10c3"}
                :body-params {:name "xxxaqq" :price 99.12}}
          resp1 (handle-update-item-by-id req1)
          resp2 (handle-update-item-by-id req2)]
      (is (and
            (= 200 (:status resp1))
            (= "pppp" (get-in resp1 [:body :name]))
            (= 23.12 (get-in resp1 [:body :price]))))
      (is (and
            (= 200 (:status resp2))
            (= "xxxaqq" (get-in resp2 [:body :name]))
            (= 99.12 (get-in resp2 [:body :price])))))))

src/inventory/core.clj
(ns inventory.core
  (:require [reitit.ring :as ring]
            [reitit.coercion.schema :as rcs]
            [muuntaja.core :as m]
            [reitit.ring.middleware.parameters :as rrmp]
            [reitit.ring.middleware.muuntaja :as rrmm]
            [reitit.ring.middleware.exception :as rrme]
            [reitit.ring.coercion :as rrc]
            [org.httpkit.server :refer [run-server]]
            [inventory.routes :refer [routes]]))

(def handler
  (ring/ring-handler
    (ring/router routes
                 {:data
                  {:coercion   rcs/coercion
                   :muuntaja   m/instance
                   :middleware [rrmp/parameters-middleware
                                rrmm/format-negotiate-middleware
                                rrmm/format-response-middleware
                                rrme/exception-middleware
                                rrmm/format-request-middleware
                                rrc/coerce-exceptions-middleware
                                rrc/coerce-request-middleware
                                rrc/coerce-response-middleware]}})))

(defonce service (atom nil))

(defn stop-service!
  []
  (when-not (nil? service)
    (@service :timeout 100)
    (reset! service nil)))

(defn start-service!
  []
  (reset! service (run-server handler {:port (int 8888)})))

(defn -main
  []
  (start-service!))

(comment
  (stop-service!)
  (-main))

src/inventory/logic.clj
(ns inventory.logic)

(defn add-item
  "Adds the item in the inventory map."
  [inventory item]
  (assoc inventory (:id item) item))

(defn remove-item
  "Removes the item of the inventory map."
  [inventory item-id]
  (dissoc inventory item-id))

(defn item-total
  "Returns the item price times the item quantity."
  [item]
  (* (:price item) (:quantity item)))

(defn item-with-total-price
  "Returns the item with the total price."
  [inventory item-id]
  (let [item (get inventory item-id)
        total (item-total item)]
    (-> item
        (assoc :total-price total))))

(defn inventory-info
  "Returns a map containing the quantity of the inventory items
   and the total value of the inventory."
  [inventory]
  (let [each-item (vals inventory)
        each-total (map item-total each-item)
        total (reduce + each-total)]
    {:inventory-items (count inventory)
     :inventory-value total}))

(defn high-amount?
  [item-quantity purchase-amount]
  (< item-quantity purchase-amount))

(defn item-purchase
  "Returns the updated inventory with the updated Item."
  [inventory purchase]
  (let [id (:item-id purchase)
        item (get inventory id)
        purchase-amount (:amount purchase)
        item-quantity (:quantity item)]
    (if (high-amount? item-quantity purchase-amount)
      {:inventory inventory :success? false}
      {:inventory (update-in inventory [id :quantity] - purchase-amount) :success? true})))

(defn item-update
  "Returns the inventory with the updated Item."
  [inventory item-id to-update]
  (update inventory item-id merge to-update))

src/inventory/handlers.clj
(ns inventory.handlers
  (:require [inventory.logic :as logic]))

(defn handle-inventory-info
  [{:keys [db]}]
  {:status 200
   :body   (logic/inventory-info @db)})

(defn handle-add-item
  [{:keys [db body-params]}]
  (let [item (assoc body-params :id (random-uuid))]
    (swap! db logic/add-item item)
    {:status 201
     :body   (logic/item-with-total-price @db (:id item))}))

(defn handle-get-item-by-id
  [{:keys [db path-params]}]
  (let [id (parse-uuid (:id path-params))]
    (if (contains? @db id)
      {:status 200 :body (logic/item-with-total-price @db id)}
      {:status 404})))

(defn handle-delete-item-by-id
  [{:keys [db path-params]}]
  (let [id (parse-uuid (:id path-params))]
    (if (not (contains? @db id))
      {:status 404}
      (do
        (swap! db logic/remove-item id)
        {:status 204}))))

(defn- item-purchase
  [db body-params]
  (let [result (logic/item-purchase @db body-params)]
    (if (not (:success? result))
      {:status 400
       :body   {:message "ERROR"}}
      (do
        (reset! db (:inventory result))
        {:status 204}))))

(defn handle-item-purchase
  [{:keys [db body-params]}]
  (let [body-params (update body-params :item-id parse-uuid)]
    (if (contains? @db (:item-id body-params))
      (item-purchase db body-params)
      {:status 400
       :body   {:message "Item not found."}})))

(defn handle-update-item-by-id
  [{:keys [db path-params body-params]}]
  (let [id (parse-uuid (:id path-params))]
    (if (contains? @db id)
      (do
        (swap! db logic/item-update id body-params)
        {:status 200
         :body   (logic/item-with-total-price @db id)})
      {:status 404})))

src/inventory/routes.clj
(ns inventory.routes
  (:require [inventory.handlers :as handlers]
            [inventory.middlewares :as middlewares]
            [schema.core :as s]
            [inventory.models :as models]))

(def empty-body (s/pred empty? 'a-empty-body))

(def routes
  [["/ok" {:name ::ok
           :get  {:handler (fn [_] {:status 200})}}]
   ["/inventory" {:name ::inventory
                  :get  {:middleware [[middlewares/db-middleware]]
                         :responses  {200 {:body {:inventory-items s/Int
                                                  :inventory-value s/Num}}}
                         :handler    handlers/handle-inventory-info}}]
   ["/items" {:name ::items
              :post {:middleware [[middlewares/db-middleware]]
                     :parameters {:body models/NewItem}
                     :responses  {201 {:body models/Item}}
                     :handler    handlers/handle-add-item}}]
   ["/items/:id" {:name   ::items-id
                  :get    {:middleware [[middlewares/db-middleware]]
                           :parameters {:path {:id s/Uuid}}
                           :responses  {200 {:body models/Item}
                                        404 {:body empty-body}}
                           :handler    handlers/handle-get-item-by-id}
                  :delete {:middleware [[middlewares/db-middleware]]
                           :parameters {:path {:id s/Uuid}}
                           :responses  {204 {:body empty-body}
                                        404 {:body empty-body}}
                           :handler    handlers/handle-delete-item-by-id}
                  :put    {:middleware [[middlewares/db-middleware]]
                           :parameters {:body models/ItemUpdate
                                        :path {:id s/Uuid}}
                           :responses  {200 {:body models/Item}}
                           :handler    handlers/handle-update-item-by-id}}]
   ["/purchase" {:name ::purchase
                 :post {:middleware [[middlewares/db-middleware]]
                        :parameters {:body {:item-id s/Uuid
                                            :amount  s/Int}}
                        :responses  {204 {:body empty-body}
                                     400 {:body {:message s/Str}}}
                        :handler    handlers/handle-item-purchase}}]])